File: leetcode/p0997/Solution.java
class Solution {
    public int findJudge(int n, int[][] trust) {
        
        List<List<Integer>> adj = new ArrayList<>();
        int[] map = new int[n + 1];
        for (int i = 0; i <= n; ++i) {
            adj.add(new ArrayList<>());
        }
        
        for (int[] t : trust) {
            adj.get(t[0]).add(t[1]);
            map[t[1]]++;
        }
        
        for (int i = 1; i <= n; ++i) {
            if (adj.get(i).isEmpty() && map[i] == n - 1) {
                return i;
            }
        }
        return -1;
    }
}

File: leetcode/p0997/Solution.kt
class Solution {
    fun findJudge(n: Int, trust: Array<IntArray>): Int {
        if (n == 1 && trust.size == 0) return 1
        
        val adj = mutableMapOf<Int, Set<Int>>()
        
        for (t in trust) {
            adj[t[1]] = ((adj[t[1]] ?: emptySet()) + t[0])
        }
        
        val judges = mutableListOf<Int>()
        for ((k, v) in adj) {
            if (v.size == n - 1) {
                judges.add(k)
            }
        }
        if (judges.size != 1) return -1
        val judge = judges.first()
        for ((k, v) in adj) {
            if (k == judge) continue
            if (v.contains(judge)) return -1
        }
        return judge
    }
}
